# Reviewing and implementing my Python Pandas

* Let's kick off!*


In [169]:
import pandas as pd
import numpy as np

In [170]:
Airfleet = ["Boeing 777", "Boeing 787", "Airbus A320", "Boeing 737", "Bombardier CRJ-700", "Embraer E170 or E175"]
fleetSeries1 = pd.Series(Airfleet) 

In [171]:
print(fleetSeries1)

0              Boeing 777
1              Boeing 787
2             Airbus A320
3              Boeing 737
4      Bombardier CRJ-700
5    Embraer E170 or E175
dtype: object


In [172]:
fleetcount = [234, 122, 337, 70, 143, 38]
fleetSeries2 = pd.Series(fleetcount) 
print(fleetSeries2)

0    234
1    122
2    337
3     70
4    143
5     38
dtype: int64


In [173]:
aircrafts = [("Boeing 777",234),("Boeing 787",122),("Airbus A320", 337),("Boeing 737", 70),("Bombardier CRJ-700", 143),("Embraer E170 or E175", 38)]
aircraftdf = pd.DataFrame(aircrafts, columns = ['Aircrafts', 'Inventory'])
aircraftdf

Aircrafts  Inventory
0            Boeing 777        234
1            Boeing 787        122
2           Airbus A320        337
3            Boeing 737         70
4    Bombardier CRJ-700        143
5  Embraer E170 or E175         38

In [174]:
aircraftdf.dtypes

Aircrafts    object
Inventory     int64
dtype: object

## Reading Crypto Data

In [175]:
mydf = pd.read_csv('/kaggle/input/top-40-crypto-perpetuals-hourly-prices/crypto.csv')

## Inspecting Data 
*Love it!*

In [176]:
mydf.head()

OpenDt  CRVUSDT-open  CRVUSDT-high  CRVUSDT-low  \
0  2020-09-04 06:00:00         2.962         2.975        2.800   
1  2020-09-04 07:00:00         2.812         2.933        2.797   
2  2020-09-04 08:00:00         2.919         3.237        2.915   
3  2020-09-04 09:00:00         3.171         3.294        3.100   
4  2020-09-04 10:00:00         3.272         3.278        3.150   

   CRVUSDT-close  CRVUSDT-volume  ZILUSDT-open  ZILUSDT-high  ZILUSDT-low  \
0          2.814      233206.032       0.01617       0.01630      0.01588   
1          2.919      422883.332       0.01594       0.01599      0.01501   
2          3.171      599623.833       0.01592       0.01654      0.01588   
3          3.282      392916.735       0.01638       0.01689      0.01621   
4          3.157      374373.881       0.01683       0.01717      0.01672   

   ZILUSDT-close  ...  EGLDUSDT-open  EGLDUSDT-high  EGLDUSDT-low  \
0        0.01592  ...         17.524         17.563        17.000   
1        0.01591  ...         17.029         17.900        16.710   
2        0.01635  ...         17.133         17.344        17.102   
3        0.01684  ...         17.277         17.277        16.918   
4        0.01703  ...         17.133         17.750        17.100   

   EGLDUSDT-close  EGLDUSDT-volume  LTCUSDT-open  LTCUSDT-high  LTCUSDT-low  \
0          17.027        13078.587         48.77         48.77        48.20   
1          17.129        24945.703         48.20         48.93        47.37   
2          17.277         7945.829         48.78         49.92        48.75   
3          17.121        15481.387         49.68         51.54        49.34   
4          17.342        19422.335         51.42         52.38        51.00   

   LTCUSDT-close  LTCUSDT-volume  
0          48.21     23686.98714  
1          48.83     46079.69323  
2          49.67     73098.31532  
3          51.45     73369.67310  
4          51.42     81565.19476  

[5 rows x 246 columns]

In [177]:
mydf.tail()

OpenDt  CRVUSDT-open  CRVUSDT-high  CRVUSDT-low  \
30308  2024-02-19 22:00:00        0.5619        0.5652       0.5601   
30309  2024-02-19 23:00:00        0.5651        0.5664       0.5599   
30310  2024-02-20 00:00:00        0.5627        0.5671       0.5561   
30311  2024-02-20 01:00:00        0.5573        0.5608       0.5489   
30312  2024-02-20 02:00:00        0.5562        0.5610       0.5554   

       CRVUSDT-close  CRVUSDT-volume  ZILUSDT-open  ZILUSDT-high  ZILUSDT-low  \
30308         0.5652        637244.0       0.02388       0.02443      0.02383   
30309         0.5628        895978.7       0.02438       0.02450      0.02419   
30310         0.5574       1362334.2       0.02433       0.02442      0.02394   
30311         0.5562       1733571.3       0.02397       0.02413      0.02366   
30312         0.5607        625111.3       0.02391       0.02426      0.02391   

       ZILUSDT-close  ...  EGLDUSDT-open  EGLDUSDT-high  EGLDUSDT-low  \
30308        0.02439  ...          60.97          61.34         60.80   
30309        0.02433  ...          61.32          61.43         60.64   
30310        0.02398  ...          60.90          61.02         59.91   
30311        0.02393  ...          60.02          60.36         59.34   
30312        0.02417  ...          59.95          61.24         59.94   

       EGLDUSDT-close  EGLDUSDT-volume  LTCUSDT-open  LTCUSDT-high  \
30308           61.32         11879.86         71.48         71.59   
30309           60.85          3605.46         71.56         71.74   
30310           60.00         10238.88         71.29         71.48   
30311           59.98          6722.68         71.00         71.22   
30312           61.17          6860.49         70.84         71.03   

       LTCUSDT-low  LTCUSDT-close  LTCUSDT-volume  
30308        71.28          71.55       17212.803  
30309        71.20          71.30       23770.375  
30310        70.70          71.00       26952.284  
30311        70.37          70.85       27091.355  
30312        70.67          71.01       16077.479  

[5 rows x 246 columns]

**We have enough data of 30312 columns hourly data from Apr 2020 to Feb 2024**
---*Openening, highs, lows, closing,volume* --- there are 246 columns which are comparing couple of cryptos stats

In [178]:
mydf.dtypes

OpenDt             object
CRVUSDT-open      float64
CRVUSDT-high      float64
CRVUSDT-low       float64
CRVUSDT-close     float64
                   ...   
LTCUSDT-open      float64
LTCUSDT-high      float64
LTCUSDT-low       float64
LTCUSDT-close     float64
LTCUSDT-volume    float64
Length: 246, dtype: object

In [179]:
mydf.describe()

CRVUSDT-open  CRVUSDT-high   CRVUSDT-low  CRVUSDT-close  \
count  30313.000000  30313.000000  30313.000000   30313.000000   
mean       1.568436      1.587214      1.548866       1.568282   
std        1.197376      1.214763      1.179621       1.197426   
min        0.332000      0.336000      0.328000       0.331000   
25%        0.644200      0.650000      0.639000       0.645000   
50%        1.028000      1.038000      1.019000       1.028000   
75%        2.264000      2.291000      2.239000       2.264000   
max        6.694000      6.800000      6.505000       6.696000   

       CRVUSDT-volume  ZILUSDT-open  ZILUSDT-high   ZILUSDT-low  \
count    3.031300e+04  30313.000000  30313.000000  30313.000000   
mean     7.783400e+05      0.055792      0.056398      0.055153   
std      1.163047e+06      0.046319      0.046994      0.045608   
min      0.000000e+00      0.013360      0.013610      0.012680   
25%      2.971420e+05      0.021950      0.022070      0.021810   
50%      5.162636e+05      0.034700      0.034980      0.034440   
75%      9.051954e+05      0.078700      0.079680      0.077750   
max      6.430103e+07      0.248560      0.256910      0.245660   

       ZILUSDT-close  ZILUSDT-volume  ...  EGLDUSDT-open  EGLDUSDT-high  \
count   30313.000000    3.031300e+04  ...   30313.000000   30313.000000   
mean        0.055792    1.520255e+07  ...      90.354493      91.240696   
std         0.046320    3.004409e+07  ...      81.059569      81.991564   
min         0.013350    0.000000e+00  ...       6.587000       6.689000   
25%         0.021950    4.642925e+06  ...      36.910000      37.100000   
50%         0.034700    8.050246e+06  ...      54.070000      54.450000   
75%         0.078690    1.493870e+07  ...     139.450000     140.740000   
max         0.248590    1.190657e+09  ...     539.420000     544.250000   

       EGLDUSDT-low  EGLDUSDT-close  EGLDUSDT-volume  LTCUSDT-open  \
count  30313.000000    30313.000000     30313.000000  30313.000000   
mean      89.425944       90.355126     12368.688953    109.160195   
std       80.066225       81.058166     19334.539700     58.361301   
min        6.000000        6.602000         0.000000     40.600000   
25%       36.690000       36.910000      3866.051800     65.700000   
50%       53.760000       54.070000      6996.940000     89.080000   
75%      138.000000      139.450000     13420.010000    142.040000   
max      528.170000      539.440000    685921.090000    409.830000   

       LTCUSDT-high   LTCUSDT-low  LTCUSDT-close  LTCUSDT-volume  
count  30313.000000  30313.000000   30313.000000    3.031300e+04  
mean     110.011576    108.246851     109.160628    3.612673e+04  
std       59.118923     57.552390      58.360284    4.509614e+04  
min       42.470000     40.300000      40.600000    0.000000e+00  
25%       65.940000     65.430000      65.700000    1.325612e+04  
50%       89.540000     88.520000      89.070000    2.233809e+04  
75%      143.390000    140.690000     142.030000    4.189789e+04  
max      413.490000    402.800000     409.840000    2.249842e+06  

[8 rows x 245 columns]

**This is my favourites helps understand data quickly**

## Accessing Row and Column

In [180]:
mydf.loc[30310]

OpenDt            2024-02-20 00:00:00
CRVUSDT-open                   0.5627
CRVUSDT-high                   0.5671
CRVUSDT-low                    0.5561
CRVUSDT-close                  0.5574
                         ...         
LTCUSDT-open                    71.29
LTCUSDT-high                    71.48
LTCUSDT-low                      70.7
LTCUSDT-close                    71.0
LTCUSDT-volume              26952.284
Name: 30310, Length: 246, dtype: object

In [181]:
mydf['LTCUSDT-open']

0        48.77
1        48.20
2        48.78
3        49.68
4        51.42
         ...  
30308    71.48
30309    71.56
30310    71.29
30311    71.00
30312    70.84
Name: LTCUSDT-open, Length: 30313, dtype: float64

Now let's try to access the row using a date instead of looking for row,
and will access that data using Feb 20th dates

In [182]:
mydf = mydf.set_index('OpenDt')
mydf

CRVUSDT-open  CRVUSDT-high  CRVUSDT-low  CRVUSDT-close  \
OpenDt                                                                        
2020-09-04 06:00:00        2.9620        2.9750       2.8000         2.8140   
2020-09-04 07:00:00        2.8120        2.9330       2.7970         2.9190   
2020-09-04 08:00:00        2.9190        3.2370       2.9150         3.1710   
2020-09-04 09:00:00        3.1710        3.2940       3.1000         3.2820   
2020-09-04 10:00:00        3.2720        3.2780       3.1500         3.1570   
...                           ...           ...          ...            ...   
2024-02-19 22:00:00        0.5619        0.5652       0.5601         0.5652   
2024-02-19 23:00:00        0.5651        0.5664       0.5599         0.5628   
2024-02-20 00:00:00        0.5627        0.5671       0.5561         0.5574   
2024-02-20 01:00:00        0.5573        0.5608       0.5489         0.5562   
2024-02-20 02:00:00        0.5562        0.5610       0.5554         0.5607   

                     CRVUSDT-volume  ZILUSDT-open  ZILUSDT-high  ZILUSDT-low  \
OpenDt                                                                         
2020-09-04 06:00:00      233206.032       0.01617       0.01630      0.01588   
2020-09-04 07:00:00      422883.332       0.01594       0.01599      0.01501   
2020-09-04 08:00:00      599623.833       0.01592       0.01654      0.01588   
2020-09-04 09:00:00      392916.735       0.01638       0.01689      0.01621   
2020-09-04 10:00:00      374373.881       0.01683       0.01717      0.01672   
...                             ...           ...           ...          ...   
2024-02-19 22:00:00      637244.000       0.02388       0.02443      0.02383   
2024-02-19 23:00:00      895978.700       0.02438       0.02450      0.02419   
2024-02-20 00:00:00     1362334.200       0.02433       0.02442      0.02394   
2024-02-20 01:00:00     1733571.300       0.02397       0.02413      0.02366   
2024-02-20 02:00:00      625111.300       0.02391       0.02426      0.02391   

                     ZILUSDT-close  ZILUSDT-volume  ...  EGLDUSDT-open  \
OpenDt                                              ...                  
2020-09-04 06:00:00        0.01592      11046278.9  ...         17.524   
2020-09-04 07:00:00        0.01591      30961740.0  ...         17.029   
2020-09-04 08:00:00        0.01635      21286989.0  ...         17.133   
2020-09-04 09:00:00        0.01684      20476272.1  ...         17.277   
2020-09-04 10:00:00        0.01703      17765957.7  ...         17.133   
...                            ...             ...  ...            ...   
2024-02-19 22:00:00        0.02439      32061081.1  ...         60.970   
2024-02-19 23:00:00        0.02433      21752675.9  ...         61.320   
2024-02-20 00:00:00        0.02398      19045265.9  ...         60.900   
2024-02-20 01:00:00        0.02393      14833038.6  ...         60.020   
2024-02-20 02:00:00        0.02417      16916191.0  ...         59.950   

                     EGLDUSDT-high  EGLDUSDT-low  EGLDUSDT-close  \
OpenDt                                                             
2020-09-04 06:00:00         17.563        17.000          17.027   
2020-09-04 07:00:00         17.900        16.710          17.129   
2020-09-04 08:00:00         17.344        17.102          17.277   
2020-09-04 09:00:00         17.277        16.918          17.121   
2020-09-04 10:00:00         17.750        17.100          17.342   
...                            ...           ...             ...   
2024-02-19 22:00:00         61.340        60.800          61.320   
2024-02-19 23:00:00         61.430        60.640          60.850   
2024-02-20 00:00:00         61.020        59.910          60.000   
2024-02-20 01:00:00         60.360        59.340          59.980   
2024-02-20 02:00:00         61.240        59.940          61.170   

                     EGLDUSDT-volume  LTCUSDT-open  LTCUSDT-high  LTCUSDT-low  \
OpenDt                 

## Subsetting Data 

In [183]:
mydf.columns

Index(['CRVUSDT-open', 'CRVUSDT-high', 'CRVUSDT-low', 'CRVUSDT-close',
       'CRVUSDT-volume', 'ZILUSDT-open', 'ZILUSDT-high', 'ZILUSDT-low',
       'ZILUSDT-close', 'ZILUSDT-volume',
       ...
       'EGLDUSDT-open', 'EGLDUSDT-high', 'EGLDUSDT-low', 'EGLDUSDT-close',
       'EGLDUSDT-volume', 'LTCUSDT-open', 'LTCUSDT-high', 'LTCUSDT-low',
       'LTCUSDT-close', 'LTCUSDT-volume'],
      dtype='object', length=245)

In [184]:
mydf = mydf[['CRVUSDT-open', 'CRVUSDT-high', 'CRVUSDT-low', 'CRVUSDT-close',
       'CRVUSDT-volume']] #altering my data to CRVUSDT only

In [185]:
mydf.shape

(30313, 5)

In [186]:
#subsetting rows based on my CRVUSDT price over 3.1
mydef_subset1 = mydf['CRVUSDT-open'] > 3.1


In [187]:
#subsetting rows of CRVUSDT price only over 5.7
mydf_subset2 = mydf.loc[mydf['CRVUSDT-open'] > 5.7]
#Query doesn't work here, Incase you like to improve this. Feel free to reach out!

In [188]:
#let's check any NaN in Open price
mydf.loc[mydf['CRVUSDT-open'].isna()]
# so here we go, no NaN values that's great


Empty DataFrame
Columns: [CRVUSDT-open, CRVUSDT-high, CRVUSDT-low, CRVUSDT-close, CRVUSDT-volume]
Index: []

In [189]:
# can use "~" here to get all the true values like 
mydf.loc[~mydf['CRVUSDT-open'].isna()]
# No effect since no NaN

CRVUSDT-open  CRVUSDT-high  CRVUSDT-low  CRVUSDT-close  \
OpenDt                                                                        
2020-09-04 06:00:00        2.9620        2.9750       2.8000         2.8140   
2020-09-04 07:00:00        2.8120        2.9330       2.7970         2.9190   
2020-09-04 08:00:00        2.9190        3.2370       2.9150         3.1710   
2020-09-04 09:00:00        3.1710        3.2940       3.1000         3.2820   
2020-09-04 10:00:00        3.2720        3.2780       3.1500         3.1570   
...                           ...           ...          ...            ...   
2024-02-19 22:00:00        0.5619        0.5652       0.5601         0.5652   
2024-02-19 23:00:00        0.5651        0.5664       0.5599         0.5628   
2024-02-20 00:00:00        0.5627        0.5671       0.5561         0.5574   
2024-02-20 01:00:00        0.5573        0.5608       0.5489         0.5562   
2024-02-20 02:00:00        0.5562        0.5610       0.5554         0.5607   

                     CRVUSDT-volume  
OpenDt                               
2020-09-04 06:00:00      233206.032  
2020-09-04 07:00:00      422883.332  
2020-09-04 08:00:00      599623.833  
2020-09-04 09:00:00      392916.735  
2020-09-04 10:00:00      374373.881  
...                             ...  
2024-02-19 22:00:00      637244.000  
2024-02-19 23:00:00      895978.700  
2024-02-20 00:00:00     1362334.200  
2024-02-20 01:00:00     1733571.300  
2024-02-20 02:00:00      625111.300  

[30313 rows x 5 columns]

## Casting Dtypes